### PR TITLE
[CI] [RL]: Add e2e test for custom pipeline 

### PR DESCRIPTION
**Pull request by knlnguyen1802**
## Purpose
This PR aim to add custom pipeline test for diffusion model which target to compatible with test on verl side.
## Test Plan
`pytest -s -v tests/e2e/offline_inference/custom_pipeline/`

cc: @SamitHuang 

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by SamitHuang** [comment]
pls supplement the test plan and result

File: tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+        height = sp.height or self.default_sample_size * self.vae_scale_factor
+        width = sp.width or self.default_sample_size * self.vae_scale_factor
+        num_inference_steps = sp.num_inference_steps or num_inference_steps
+        max_sequence_length = sp.max_sequence_length or max_sequence_length
+
+        noise_level = sp.extra_args.get("noise_level", None) or noise_level
+        sde_window_size = sp.extra_args.get("sde_window_size", None) or sde_window_size
+        sde_window_range = sp.extra_args.get("sde_window_range", None) or sde_window_range
+        sde_type = sp.extra_args.get("sde_type", None) or sde_type
+        logprobs = sp.extra_args.get("logprobs", None)
```
Comment (SamitHuang) [inline tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py]:
```suggestion
        logprobs = sp.extra_args.get("logprobs", None) or logprobs
```

File: tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+                noise_norm = torch.norm(comb_pred, dim=-1, keepdim=True)
+                noise_pred = comb_pred * (cond_norm / noise_norm)
+            # compute the previous noisy sample x_t -> x_t-1
+            latents, log_prob, _, _ = self.scheduler.step(
+                noise_pred,
+                t,
+                latents,
+                generator=generator,
+                noise_level=cur_noise_level,
+                sde_type=sde_type,
```
Comment (SamitHuang) [inline tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py:190]:
```suggestion
                sde_type=sde_type,
                return_logprobs=logprobs,
```

**Review by lishunyang12 — COMMENTED** [review]
Left a few comments, mostly around the scheduler and the logprobs param handling.

File: tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+        height = sp.height or self.default_sample_size * self.vae_scale_factor
+        width = sp.width or self.default_sample_size * self.vae_scale_factor
+        num_inference_steps = sp.num_inference_steps or num_inference_steps
+        max_sequence_length = sp.max_sequence_length or max_sequence_length
+
+        noise_level = sp.extra_args.get("noise_level", None) or noise_level
+        sde_window_size = sp.extra_args.get("sde_window_size", None) or sde_window_size
+        sde_window_range = sp.extra_args.get("sde_window_range", None) or sde_window_range
+        sde_type = sp.extra_args.get("sde_type", None) or sde_type
+        logprobs = sp.extra_args.get("logprobs", None)
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py]: When `"logprobs"` is absent from `extra_args`, this evaluates to `None` and silently overrides the function default (`logprobs=True`). The surrounding lines use `or` to fall back, but this one doesn't:

```suggestion
        logprobs = sp.extra_args.get("logprobs", logprobs)
```

File: tests/e2e/offline_inference/custom_pipeline/flow_match_sde_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                The noise level used in the SDE.
+            prev_sample (`torch.FloatTensor`, *optional*):
+                The sample from the previous timestep. If not provided, it will be sampled inside the function.
+            sde_type (`str`, *optional*, defaults to "sde"):
+                The type of SDE to use. Choose between "sde" and "cps".
+            return_logprobs (`bool`, *optional*, defaults to True):
+                Whether to return log probabilities of the previous sample.
+        """
+
+        if isinstance(timestep, int) or isinstance(timestep, torch.IntTensor) or isinstance(timestep, torch.LongTensor):
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/custom_pipeline/flow_match_sde_scheduler.py]:
```suggestion
        if isinstance(timestep, (int, torch.IntTensor, torch.LongTensor)):
```

File: tests/e2e/offline_inference/custom_pipeline/flow_match_sde_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+        timestep: torch.FloatTensor | None = None,
+        generator: torch.Generator | None = None,
+        per_token_timesteps: torch.Tensor | None = None,
+        noise_level: float = 0.7,
+        prev_sample: torch.Tensor | None = None,
+        sde_type: Literal["cps", "sde"] = "sde",
+        return_logprobs: bool = True,
+    ):
+        assert sde_type in ["sde", "cps"]
+        assert sample.dtype == torch.float32
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/custom_pipeline/flow_match_sde_scheduler.py:151]: Bare `assert` is stripped under `python -O`. Since this is a scheduler (not a test file), use `if … raise ValueError` instead.

File: tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    if isinstance(v, torch.Tensor):
+        return v.detach().cpu()
+    return v
+
+
+# Custom pipeline class for QwenImage that returns log probabilities during the diffusion process.
+# This is for test
+class QwenImagePipelineWithLogProbForTest(QwenImagePipeline):
+    def __init__(self, *, od_config: OmniDiffusionConfig, prefix: str = ""):
+        super(QwenImagePipeline, self).__init__()
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py]: Explicitly calling grandparent `__init__` via `super(QwenImagePipeline, self)` is fragile — it'll silently break if the MRO changes. Worth a `# skip QwenImagePipeline.__init__ because …` comment so future readers know this is intentional.

File: tests/e2e/offline_inference/custom_pipeline/test_async_omni_qwen_image_generate.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        try:
+            normalized_ids.append(int(token_id))
+        except (TypeError, ValueError) as e:
+            raise TypeError(f"token_id must be int-convertible, got {type(token_id).__name__}: {token_id!r}") from e
+    return normalized_ids
+
+
+def _tokenize_prompt(text: str) -> list[int]:
+    """Tokenize a text prompt into valid token IDs for the model."""
+    tokenizer = AutoTokenizer.from_pretrained(TOKENIZER_MODEL, trust_remote_code=True)
```
Comment (lishunyang12) [inline tests/e2e/offline_inference/custom_pipeline/test_async_omni_qwen_image_generate.py:82]: This downloads and instantiates a tokenizer on every call. In `test_async_omni_generate_concurrent` it runs 4× in sequence. Consider a module-level `@lru_cache` or a `@pytest.fixture(scope="module")` to avoid redundant downloads.

**Review by SamitHuang — APPROVED** [review]
LGTM

File: tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+        height = sp.height or self.default_sample_size * self.vae_scale_factor
+        width = sp.width or self.default_sample_size * self.vae_scale_factor
+        num_inference_steps = sp.num_inference_steps or num_inference_steps
+        max_sequence_length = sp.max_sequence_length or max_sequence_length
+
+        noise_level = sp.extra_args.get("noise_level", None) or noise_level
+        sde_window_size = sp.extra_args.get("sde_window_size", None) or sde_window_size
+        sde_window_range = sp.extra_args.get("sde_window_range", None) or sde_window_range
+        sde_type = sp.extra_args.get("sde_type", None) or sde_type
+        logprobs = sp.extra_args.get("logprobs", None) or logprobs
```
Comment (zhtmike) [inline tests/e2e/offline_inference/custom_pipeline/qwen_image_pipeline_with_logprob.py:260]: default logprobs: bool = True

so `logprobs = sp.extra_args.get("logprobs", None) or logprobs` is always True.
